syntax = "proto3";

message SetupConnection {

    message Cluster {
        string name = 1;
        string ip = 2;
        uint32 port = 3;
    }

    repeated Cluster all_clusters = 2;
}

message StartElection {
    string src = 1;
    uint32 term = 2;
}

message RequestVote {
    uint32 term = 1;
    string clusterName = 2;
    string clusterIp = 3;
    uint32 clusterPort = 4;
}

message ResponseVote {
    uint32 term = 1;
    string voteStatus = 2;
    string clusterName = 3;
}

message HeartBeatMessage {
    uint32 term = 1;
    string clusterName = 2;
}

message KeyValueMessage {
    oneof key_value_message {
        SetupConnection setup_connection = 1;
        StartElection start_election = 2;
        RequestVote request_vote = 3;
        ResponseVote response_vote = 4;
        HeartBeatMessage heartbeat_message = 5;
    }
}
